server/routes/user_router.go

```diff
@@ -0,0 +1,358 @@
+		return
+	}
+	if !tkn.Valid {
+		router.WriteErrResponse(w, fmt.Errorf("token is invalid"))
+		return
+	}
+
+	if time.Until(claims.ExpiresAt.Time) > 30*time.Minute {
+		router.WriteErrResponse(w, fmt.Errorf("token is expired"))
+		return
+	}
+
+	expirationTime := time.Now().Add(5 * time.Minute)
+	claims.ExpiresAt = jwt.NewNumericDate(expirationTime)
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+	newToken, err := token.SignedString([]byte(router.secret))
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+	router.WriteMsgResponse(w, "Old Token: "+reqToken+"\n New Token: "+newToken)
+}
+
+func (router *Router) Logout(w http.ResponseWriter, r *http.Request) {
+	reqToken := r.Header.Get("Authorization")
+	splitToken := strings.Split(reqToken, "Bearer ")
+	reqToken = splitToken[1]
+
+	claims := &models.Claims{}
+	tkn, err := jwt.ParseWithClaims(reqToken, claims, func(token *jwt.Token) (interface{}, error) {
+		return []byte(router.secret), nil
+	})
+
+	if err != nil {
+		if err == jwt.ErrSignatureInvalid {
+			router.WriteErrResponse(w, err)
+			return
+		}
+		router.WriteErrResponse(w, err)
+		return
+	}
+	if !tkn.Valid {
+		router.WriteErrResponse(w, fmt.Errorf("token is already invalid"))
+		return
+	}
+
+	expirationTime := time.Now()
+	claims.ExpiresAt = jwt.NewNumericDate(expirationTime)
+	router.WriteMsgResponse(w, "Logged out successfully")
+
+}
+
+func (router *Router) ForgotPasswordHandler(w http.ResponseWriter, r *http.Request) {
+	var email string
+	err := json.NewDecoder(r.Body).Decode(&email)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	// send verification code
+	code, err := internal.SendMail(email)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+	fmt.Printf("code: %v\n", code)
+
+	msg := "Verfification Code has been sent to " + email
+
+	router.db.SetCache(email, code)
+	router.WriteMsgResponse(w, msg)
+}
+
+func (router *Router) VerifyCode(w http.ResponseWriter, r *http.Request) {
+	data := AuthDataInput{}
+	err := json.NewDecoder(r.Body).Decode(&data)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	cachedUser, err := router.db.GetCache(data.Email)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	if cachedUser.Code != data.Code {
+		router.WriteErrResponse(w, fmt.Errorf("wrong code"))
+		return
+	}
+
+	msg := "Code Verified"
+	msgBytes, err := json.Marshal(msg)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+	}
+	router.WriteMsgResponse(w, msgBytes)
+}
+
+func (router *Router) ChangePassword(w http.ResponseWriter, r *http.Request) {
+	data := ChangePasswordInput{}
+	err := json.NewDecoder(r.Body).Decode(&data)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	err = router.db.UpdatePassword(data.Email, data.Password)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+}
+
+func (router *Router) UpdateAccount(w http.ResponseWriter, r *http.Request) {
```

Comment:
-> UpdateUserHandler
- it should take should take a user ID in the api params -> /update../{id}
note that update user can send only one field (name only for example) to update, so we need 
- an updateUserInput struct that has all fields that could be updated optional 
- validate it
- update the fields that the user has sent only